spi/max7219_8x7seg_spi/max7219_8x7seg_spi.c

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright (c) 2022 Raspberry Pi (Trading) Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include "pico/stdlib.h"
+#include "pico/binary_info.h"
+#include "hardware/spi.h"
+
+/* Example code to talk to a Max7219 driving an 8 digit 7 segment display via SPI
+
+   NOTE: The device is driven at 5v, but SPI communications are at 3v3
+
+   Connections on Raspberry Pi Pico board and a generic Max7219 board, other
+   boards may vary.
+
+   * GPIO 17 (pin 22) Chip select -> CS on max7219 board
+   * GPIO 18 (pin 24) SCK/spi0_sclk -> CLK on max7219 board
+   * GPIO 19 (pin 25) MOSI/spi0_tx -> DIN on max7219 board
+   * 3.3v (pin 36) -> VCC on max7219 board
+   * GND (pin 38)  -> GND on max7219 board
+
+   Note: SPI devices can have a number of different naming schemes for pins. See
+   the Wikipedia page at https://en.wikipedia.org/wiki/Serial_Peripheral_Interface
+   for variations.
+
+*/
+
+// This defines how many Max7219 modules we have cascaded together, in this case, just the one.
+#define NUM_MODULES 1
+
+const uint8_t CMD_NOOP = 0;
+const uint8_t CMD_DIGIT0 = 1; // Goes up to 8, for each line
+const uint8_t CMD_DECODEMODE = 9;
+const uint8_t CMD_BRIGHTNESS = 10;
+const uint8_t CMD_SCANLIMIT = 11;
+const uint8_t CMD_SHUTDOWN = 12;
+const uint8_t CMD_DISPLAYTEST = 15;
+
+#ifdef PICO_DEFAULT_SPI_CSN_PIN
+static inline void cs_select() {
+    asm volatile("nop \n nop \n nop");
+    gpio_put(PICO_DEFAULT_SPI_CSN_PIN, 0);  // Active low
+    asm volatile("nop \n nop \n nop");
+}
+
+static inline void cs_deselect() {
+    asm volatile("nop \n nop \n nop");
+    gpio_put(PICO_DEFAULT_SPI_CSN_PIN, 1);
+    asm volatile("nop \n nop \n nop");
+}
+#endif
+
+#if defined(spi_default) && defined(PICO_DEFAULT_SPI_CSN_PIN)
+static void write_register(uint8_t reg, uint8_t data) {
+    uint8_t buf[2];
+    buf[0] = reg;
+    buf[1] = data;
+    cs_select();
+    spi_write_blocking(spi_default, buf, 2);
+    cs_deselect();
+    sleep_ms(1);
+}
+
+static void write_register_all(uint8_t reg, uint8_t data) {
+    uint8_t buf[2];
+    buf[0] = reg;
+    buf[1] = data;
+    cs_select();
+    for (int i = 0; i< NUM_MODULES;i++)
+        spi_write_blocking(spi_default, buf, 2);
+    cs_deselect();
+ 
+
```

Comment:
Ahhh, this seems to be where the closing-brace is missing?

Reply:
I don't understand why this was working at my end!! Perhaps a typo crept in after testing. Will sort it all out Monday.